src/main/host/process.rs

```diff
@@ -579,6 +580,88 @@ impl Process {
+        // any other threads left alive in the process, perform a futex wake on
+        // that address. This mechanism is typically used in `pthread_join` etc.
+        // See `set_tid_address(2)`.
+        let clear_child_tid_pvp = unsafe { cshadow::thread_getTidAddress(thread.cthread()) };
+        if clear_child_tid_pvp.val != 0 && self.threads.borrow().len() > 1 && !self.is_exiting.get()
+        {
+            // Wait until the thread is really dead. It might not be dead yet if we
+            // marked the thread dead after seeing the `exit` syscall - the managed
+            // process might not have finished executing the native syscall yet.
+            // TODO: Move this into the `exit` syscall handling and/or use the native
+            // tidaddress mechanism to be notified when a thread has actually terminated.
+            let pid = unsafe { cshadow::thread_getNativePid(thread.cthread()) };
+            let tid = unsafe { cshadow::thread_getNativeTid(thread.cthread()) };
+            loop {
+                match tgkill(pid, tid, 0) {
+                    Err(Errno::ESRCH) => {
+                        trace!("Thread is done exiting; proceeding with cleanup");
+                        break;
+                    }
+                    Err(e) => {
+                        error!("Unexpected tgkill error: {:?}", e);
+                        break;
+                    }
+                    Ok(()) if pid == tid => {
+                        // Thread leader could be in a zombie state waiting for
+                        // the other threads to exit.
+                        let filename = format!("/proc/{pid}/stat");
+                        let stat = match std::fs::read_to_string(filename) {
+                            Err(e) => {
+                                assert!(e.kind() == std::io::ErrorKind::NotFound);
+                                trace!("tgl {pid} is fully dead");
+                                break;
+                            }
+                            Ok(s) => s,
+                        };
+                        if stat.contains(") Z") {
+                            trace!("tgl {pid} is a zombie");
+                            break;
+                        }
+                        // Still alive and in a non-zombie state; continue
+                    }
+                    Ok(()) => {
+                        // Thread is still alive; continue.
+                    }
+                };
+                debug!("{pid}.{tid} still running; waiting for it to exit");
+                std::thread::yield_now();
+                // Check again
+            }
+
+            // Find a still-living thread to execute the memory-write.
+            let threads = self.threads.borrow();
+            let writer_thread = threads
+                .values()
+                .find(|t| unsafe { cshadow::thread_isRunning(t.cthread()) })
+                .unwrap();
```

Comment:
> Could you just use the process group leader pid instead (or maybe just the process pid?), which I think should always live at least as long as any other thread?

An earlier implementation did that, but it turns out it doesn't always work. While the process group leader will still exist after exiting if other threads are still alive, it'll be in a zombie state and doesn't work with `process_vm_writev`.
https://github.com/shadow/shadow/blob/main/src/main/host/memory_manager/memory_copier.rs#L210

> but does that guarantee that there is a thread running? This seems like it could panic?

Yeah, there shouldn't be any non-running threads in the table. Simplified and documented this a bit more.